### DISCUSSION TITLE
glog redefinition using ceres solver and or-tools together

### DISCUSSION THREAD

**Discussion by Youhaochan** [google/or-tools#4165] [Build and configuration issues]
My C++ project currently utilizes the VRP algorithm implemented in OR-Tools, and the NLP solver in Ceres Solver. However, when attempting to build my project, I encountered the following error: 

**/usr/local/include/glog/logging.h:675:8: error: redefinition of ‘struct google::CheckOpString’
  675 | struct CheckOpString {
      |        ^~~~~~~~~~~~~
/usr/local/include/ortools/base/logging.h:504:8: note: previous definition of ‘struct google::CheckOpString’
  504 | struct CheckOpString {
      |        ^~~~~~~~~~~~~**

It appears that OR-Tools has its own Glog file, and Ceres relies on the external Glog library, resulting in a conflict. How can I resolve this issue?

**Comment by Mizux** [comment]
First, GLOG is deprecated (and no more maintained by Google)...

In v9.5, code was moved to `ortools/base/log.h`
https://github.com/google/or-tools/blob/92eab40155d5566039c92662632f196174f8da47/ortools/base/log.h#L455

In v9.6, it was replaced by abs::log (finally released after few years of use internally as replacement of GLOG...)
see: https://github.com/google/or-tools/blob/main/ortools/base/logging.h

Which version of or-tools did you use ?

**Reply by Youhaochan** [reply]
Previously, I was using version 9.0 of OR-Tools. Upon encountering a conflict, I attempted to upgrade to version 9.6 as per your suggestion, where I learned that 'logging.h' had been replaced by 'absl log' since GLog. Despite this update, I continued to encounter a redefinition error. It appears that there may be compatibility issues between Ceres Solver (v2.2) and OR-Tools when used together. The specific error message is detailed below.
**/usr/local/include/glog/log_severity.h:53:11: note: previous declaration ‘const int google::GLOG_INFO’
   53 | const int GLOG_INFO = 0, GLOG_WARNING = 1, GLOG_ERROR = 2, GLOG_FATAL = 3,
      |           ^~~~~~~~~
In file included from /usr/local/include/ortools/constraint_solver/routing.h:177,
/usr/local/include/ortools/base/logging.h:49:62: error: ‘GLOG_WARNING’ redeclared as different kind of entity
   49 |   GLOG_WARNING = static_cast<int>(absl::LogSeverity::kWarning),**

**/usr/local/include/glog/logging.h:489:1: note: previous declaration ‘google::int32 fLI::FLAGS_v’
  489 | DECLARE_int32(v);  // in vlog_is_on.cc
      | ^~~~~~~~~~~~~
In file included from /usr/local/include/ortools/base/logging.h:22,
/usr/local/include/ortools/base/vlog_is_on.h:21:1: error: ‘absl::lts_20230125::Flag<int> FLAGS_v’ conflicts with a previous declaration
   21 | ABSL_DECLARE_FLAG(int, v);**

**Reply by Mizux** [reply]
https://github.com/google/or-tools/commit/0ddf9d597896a57e203d746dbc8c7ae2956b3351
seems to have been removed for v9.9 ...
